src/libopenrave/kinbody.cpp

```diff
@@ -5348,56 +5446,141 @@ const std::vector<int>& KinBody::GetNonAdjacentLinks(int adjacentoptions) const
-            boost::static_pointer_cast<Link>(_veclinks[i])->_info._t = _vInitialLinkTransformations.at(i);
+            boost::static_pointer_cast<Link>(_veclinks[i])->_info._t = positionConfigurationAndLinkTransformations.second.at(i);
         }
         _nUpdateStampId++; // because transforms were modified
-        _vNonAdjacentLinks[0].resize(0);
 
-        for(size_t ind0 = 0; ind0 < _veclinks.size(); ++ind0) {
-            for(size_t ind1 = ind0+1; ind1 < _veclinks.size(); ++ind1) {
-                const bool bAdjacent = AreAdjacentLinks(ind0, ind1);
-                if(!bAdjacent && !collisionchecker->CheckCollision(LinkConstPtr(_veclinks[ind0]), LinkConstPtr(_veclinks[ind1])) ) {
-                    _vNonAdjacentLinks[0].push_back(ind0|(ind1<<16));
+        // Constructs initial dofValueIsDeterminableFlags
+        std::fill(dofValueIsDeterminableFlags.begin(), dofValueIsDeterminableFlags.end(), false);
+        for( const PositionConfiguration::JointConfigurationState& jointConfiguration : positionConfigurationAndLinkTransformations.first->jointConfigurationStates ) {
+            JointPtr joint = GetJoint(jointConfiguration.GetResolvedJointName());
+            if( !!joint && jointConfiguration.jointAxis < joint->GetDOF() ) {
+                dofValueIsDeterminableFlags.at(joint->GetDOFIndex() + jointConfiguration.jointAxis) = true;
+            }
+        }
+
+        // Looks up which DOF values are determinable given the position configuration
+        static constexpr int maxNumIterations = 5;
+        if( !_ResolveDOFValueDeterminableFlags(dofValueIsDeterminableFlags, maxNumIterations) ) {
+            RAVELOG_WARN_FORMAT("Failed to calculate a converged list of value-determinable DOFs within %d attempts."
+                                " There might be a circular mimic DOF dependencies or too deep mimic DOF dependencies."
+                                " Continuing with a possibly subset of value-determinable DOFs, which is on the safe side."
+                                " (i.e. there might be robot links on which self-collision checking is performed unintentionally.)", maxNumIterations);
+        }
+        const bool areAllDOFValuesDeterminable = std::find(dofValueIsDeterminableFlags.begin(), dofValueIsDeterminableFlags.end(), false) == dofValueIsDeterminableFlags.end();
+
+        // Updates adjacentLinkFlags
+        for( size_t ind0 = 0; ind0 < _veclinks.size(); ++ind0 ) {
+            for( size_t ind1 = ind0 + 1; ind1 < _veclinks.size(); ++ind1 ) {
+                const int adjacentLinkFlagIndex = _GetIndex1d(ind0, ind1);
+                if( adjacentLinkFlags.at(adjacentLinkFlagIndex) ) {
+                    continue;  // Already marked as adjacent, no need to check again
+                }
+                if( AreAdjacentLinks(ind0, ind1) ) {
+                    adjacentLinkFlags.at(adjacentLinkFlagIndex) = true;  // Optimisation; allows skipping the check next time
```

Comment:
If I understand correctly, `adjacentLinkFlags` in this function has the same size and format as `_vAdjacentLinks` which you check against here via `AreAdjacentLinks`. In that case, why don't we initialize `adjacentLinkFlags` to `_vAdjacentLinks` in line 5455 and skip this comparison inside the loop?

Reply:
Those are not equivalent. `_vAdjacentLinks` is calculated from actual link adjacencies and user-specified `KinBodyInfo::_vForcedAdjacentLinks`, while this local variable is calculated from user-specified non-self-colliding configurations. Please see the implementation of `KinBody::_ComputeInternalInformation()`.

Reply:
I had a quick look at it and will read in detail tomorrow, but I don't think the calculation of `_vAdjacentLink` is central to this question: Since you perform the check in this line, doesn't `adjacentLinkFlags` become a superset of `_vAdjacentLink` (which is a superset of `_vForcedAdjacentLinks`)? If yes, shouldn't we just initialize it to that value?

Reply:
Sorry, I was misunderstanding your point. I updated the code to do so.